hw/ip/otbn/dv/rig/rig/gens/bad_bnmovr.py

```diff
@@ -0,0 +1,111 @@
+
+
+class BadBNMovr(SnippetGen):
+    '''A snippet generator that generates program ending BN.MOVR instructions.
+
+    This includes incrementing both GPRs or having *grd or *grs > 31
+
+    '''
+
+    ends_program = True
+
+    def __init__(self, cfg: Config, insns_file: InsnsFile) -> None:
+        super().__init__()
+
+        self.insn = self._get_named_insn(insns_file, 'bn.movr')
+
+        # bn.movr expects the operands: grd, grs, grd_inc, grs_inc
+        if len(self.insn.operands) != 4:
+            raise RuntimeError('Unexpected number of operands for bn.movr')
+
+        grd, grs, grd_inc, grs_inc = self.insn.operands
+        exp_shape = (
+            # grd
+            isinstance(grd.op_type, RegOperandType) and
+            grd.op_type.reg_type == 'gpr' and
+            not grd.op_type.is_dest() and
+            # grs
+            isinstance(grs.op_type, RegOperandType) and
+            grs.op_type.reg_type == 'gpr' and
+            not grs.op_type.is_dest() and
+            # grd_inc
+            isinstance(grd_inc.op_type, OptionOperandType) and
+            # grs_inc
+            isinstance(grs_inc.op_type, OptionOperandType)
+        )
+        if not exp_shape:
+            raise RuntimeError('Unexpected shape for bn.movr')
+
+        self.weight = cfg.insn_weights.get(self.insn.mnemonic)
+
+        # Check that the instruction has a positive weight
+        if not self.weight:
+            self.disabled = True
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+
+        # Get known registers
+        known_regs = model.regs_with_known_vals('gpr')
+        bad_regs = []
+        good_regs = []
+
+        # Make sure there is at least one out of bounds value
+        for reg_idx, reg_val in known_regs:
+            if 31 < reg_val:
+                bad_regs.append(reg_idx)
+            if reg_val <= 31:
+                good_regs.append(reg_idx)
+
+        # We have 4 different options for increments * 4 different options for
+        # out of bound register values.
+        # reg_val_choice[1:0] ->   grd_val, grs_val
+        # inc_choice[1:0]     ->   grd_inc, grs_inc
+        # bad = 1, good = 0
+        # choices[3:0] = {reg_val_choice, inc_choice}
+
+        # We always have an element (x0) in good_regs but if bad_regs is empty
+        # hard-code grd_val/grs_val = Good/Good and we have to use two incr.
+        # for producing an error.
+        choices = 3 if not bad_regs else random.randint(3, 15)
+
+        reg_val_choice = choices // 4
+        inc_choice = choices % 4
+
+        bad_grd = reg_val_choice in [2, 3]  # grd_val/grs_val = Bad/Good or Bad/Bad
+        bad_grs = reg_val_choice in [1, 3]  # grd_val/grs_val = Good/Bad or Bad/Bad
+
+        # If grs_val/grd_val = good/good, we have to force +/+ for a fault
+        grd_inc = inc_choice in [2, 3] or not reg_val_choice
+        grs_inc = inc_choice in [1, 3] or not reg_val_choice
```

Comment:
I don't think you need the `or not reg_val_choice` bit, because you're picking `choices` to be at least 3, right? Also, I'm slightly surprised mypy doesn't complain about this: I think that `grd_inc` and `grs_inc` are bools here, but we expect integers in `op_val`. Maybe wrap with `int(..)` to convert explicitly?